### PR TITLE
Add Alpha architecture

### PR DESCRIPTION
**Pull request by R33v0LT**
This PR adds Alpha (auto-sync) architecture for capstone

**New Features**
- capstone support for alpha architecture
- alpha python bindings
- new tests (c/python)
- alpha support in Updater (auto-sync)

### CODE REVIEW

File: arch/Alpha/AlphaDisassembler.c
Comment (Rot127) [inline arch/Alpha/AlphaDisassembler.c]: There is a function for this in `utils.c`

File: arch/Alpha/AlphaInstPrinter.c
Comment (Rot127) [inline arch/Alpha/AlphaInstPrinter.c]: Please call this function `Alpha_LLVM_printInst(...)` and define it in a separated header `AlphaLinkage.h`. And do the same for the other static LLVM functions which are called from the `AlphaMapper` component (`getInstruction`, `getRegisterName`).

`Alpha_LLVM_printInst` should be called by `AlphaMapping.c::Alpha_printer()`. `Alpha_printer()` should be set as printer in `AlphaModule`.

So the call path is:
`CS` -> `csh->printer()` = `AlphaMapping.c::Alpha_printer()` -> `AlphaInstPrinter.c::Alpha_LLVM_printInst()` -> `AlphaInstPrinter.c::printInst()`

The point of this is that we have only a single function (`Alpha_LLVM_printInst()`) which does nothing else but link the Capstone to the LLVM code. And if you have to manipulate anything on the LLVM data, you can do it in `AlphaMapping.c::Alpha_printer()`. This keeps the `AlphaInstPrinter.c` file relatively close to its C++ equivalent.
And we still have a place to add our code (in `AlphaMapping.c::Alpha_printer()`).

You can check out how it is done in `ARM` and the other auto-sync archs.

File: arch/Alpha/AlphaMapping.c
Comment (Rot127) [inline arch/Alpha/AlphaMapping.c]: Don't forget to checkout which of those mapping functions are already implemented in `Mapping.c`.

File: bindings/python/test_alpha.py
Comment (Rot127) [inline bindings/python/test_alpha.py]: `python` -> `python3` pls

File: include/capstone/alpha.h
Comment (Rot127) [inline include/capstone/alpha.h:5]: Set your copyright here.

File: include/capstone/alpha.h
Comment (Rot127) [inline include/capstone/alpha.h]: Move the `8` to its own `#define MAX_ALPHA_OPS 8`. Will be useful in the future.

**Comment by Rot127** [comment]
Looks good so far. Nice work!

**Review by XVilka — CHANGES_REQUESTED** [review]
Don't forget to add yourself in CREDITS.TXT

File: arch/Alpha/AlphaDisassembler.c
Comment (R33v0LT) [reply arch/Alpha/AlphaDisassembler.c]: changed to `readBytes` from `utils.c`

File: arch/Alpha/AlphaInstPrinter.c
Comment (R33v0LT) [reply arch/Alpha/AlphaInstPrinter.c]: Done for `Alpha_getRegisterName` and `Alpha_printInst`

File: include/capstone/alpha.h
Comment (R33v0LT) [reply include/capstone/alpha.h]: That was a mistake. There is 3 max operands. Changed it and add `#define MAX_ALPHA_OPS 3`

**Comment by XVilka** [comment]
@R33v0LT since ARM was merged into `next`, and @Rot127 updated PPC, please update this PR too.

File: arch/Alpha/AlphaInstPrinter.c
Comment (Rot127) [inline arch/Alpha/AlphaInstPrinter.c]: Please checkout `Mapping.*`. We have a lot of helper functions for those checks and member access stuff (like `detail_is_set(MI)` or `<ARCH>_get_detail()`, `get_detail_op` etc.)

File: arch/Alpha/AlphaInstPrinter.c
Comment (Rot127) [inline arch/Alpha/AlphaInstPrinter.c]: Move those kind of `detail` filling functions to `AlphaMapping.*` please. So they follow the principle design of all `auto-sync` archs.

Comment (Rot127) [reply arch/Alpha/AlphaInstPrinter.c]: Also implement the signature just as in PPC or ARM. For example you need to pass the `OpNum`, because in `AlphaGenCSMappingInsnOp.inc` everything is indexed by the LLVM `OpNum`

File: arch/Alpha/AlphaGenCSMappingInsnOp.inc
Comment (Rot127) [inline arch/Alpha/AlphaGenCSMappingInsnOp.inc]: The reason you have `CS_OP_INVALID` here is https://github.com/capstone-engine/llvm-capstone/pull/17/files#r1281777199

**Comment by Rot127** [comment]
Generally looks good so far. Please don't forget to set the details of each operand (read/write access and the like).
You can look at [the PPC implementation](https://github.com/Rot127/capstone/blob/028e4e35365d4a96170ffdd98c0100cd1eb985d9/arch/PowerPC/PPCMapping.c#L482-L538) for an example

**Comment by XVilka** [comment]
@R33v0LT please rebase on top of the latest `next` and `aarch64` PR

**Comment by XVilka** [comment]
@R33v0LT I wonder why there are still commits by @Rot127 - they should already be part of `next`. Maybe you could just cherry-pick your commits, and push again?

**Comment by Rot127** [comment]
@R33v0LT You need to rebase again. @kabeor purged the `test_aarch64` binary from the commit history. So it was rewritten, this is why you've got +2k commits here

**Comment by XVilka** [comment]
@R33v0LT seems there are still unnecessary commits

**Comment by XVilka** [comment]
@Rot127, please take a look again. @kabeor could you please allow workflows to run?

@R33v0LT What is still missing? Why this is still a draft?

**Comment by XVilka** [comment]
```
Traceback (most recent call last):
  File "/home/runner/work/capstone/capstone/bindings/python/./test_mos65xx.py", line 95, in <module>
    test_class()
  File "/home/runner/work/capstone/capstone/bindings/python/./test_mos65xx.py", line 86, in test_class
    print_insn_detail(insn)
  File "/home/runner/work/capstone/capstone/bindings/python/./test_mos65xx.py", line 58, in print_insn_detail
    print("\taddress mode: %s" % (address_modes[insn.am]))
IndexError: list index out of range
FAILED
make: *** [Makefile:78: check] Error 1
Error: Process completed with exit code
```
Also:
```
Test MC/ARM/neont2-mul-accum-encoding.s.cs
TeTraceback (most recent call last):
  File "./test_corpus.py", line 155, in <module>
    test_file(fname.strip())
  File "./test_corpus.py", line 146, in test_file
    fout.write(unichr(mc_modes[(arch, mode)]))
KeyError: ('CS_ARCH_ALPHA', '0')
st MC/ARM/neon-satshift-encoding.s.cs
```

**Comment by XVilka** [comment]
@kabeor could you please approve workflows to run?
@R33v0LT why removing it from fuzzing? Technically it should be crash-free too.

**Comment by R33v0LT** [comment]
> @kabeor could you please approve workflows to run? @R33v0LT why removing it from fuzzing? Technically it should be crash-free too.

I mistakenly thought that fuzzing tests were related to llvm-mc. Fixed it

**Comment by R33v0LT** [comment]
> @Rot127, please take a look again. @kabeor could you please allow workflows to run?
> 
> @R33v0LT What is still missing? Why this is still a draft?

I can't figure out why python bindings are broken. As soon as I figure it out, I'll remove it from the draft

**Comment by XVilka** [comment]
@R33v0LT I suppose it's not a draft anymore. Please remove the mark also update the PR description. @Rot127, please take a look again.

File: .gitmodules
Comment (Rot127) [inline .gitmodules:1]: Please don't remove the sub-module

File: arch/Alpha/AlphaGenCSFeatureName.inc
Comment (Rot127) [inline arch/Alpha/AlphaGenCSFeatureName.inc:1]: You can remove empty files.

File: include/capstone/alpha.h
Comment (Rot127) [inline include/capstone/alpha.h]: Also here. If it is not used, please remove it.

File: cstool/cstool.c
Comment (Rot127) [inline cstool/cstool.c:117]: Please check if you want to replace `0` with `CS_MODE_LITTLE_ENDIAN`

File: bindings/python/test_alpha.py
Comment (Rot127) [reply bindings/python/test_alpha.py]: Please don't forget about this one.

File: arch/Alpha/AlphaMapping.c
Comment (Rot127) [inline arch/Alpha/AlphaMapping.c]: You do `if (i == 0) { return; }`. Makes it the following more readable because less indentations.

File: arch/Alpha/AlphaGenCSFeatureName.inc
Comment (Rot127) [inline arch/Alpha/AlphaGenCSFeatureName.inc:1]: Same as for the others unused. Please remove.

File: suite/cstest/src/alpha_detail.c
Comment (Rot127) [inline suite/cstest/src/alpha_detail.c:4]: Your copyright header please.

File: tests/test_alpha.c
Comment (Rot127) [inline tests/test_alpha.c:4]: Copyright

File: suite/auto-sync/Updater/Update-Arch.sh
Comment (Rot127) [inline suite/auto-sync/Updater/Update-Arch.sh:1]: See review comment

**Review by Rot127 — CHANGES_REQUESTED** [review]
Thanks a lot! It looks very good!

Besides the few comments from above, please do the following before it can be merged:

**Fuzz it a little more:**

Apply the patch from [here](https://github.com/capstone-engine/capstone/pull/2212#issue-2019456045) and test every byte combination for segfaults by running:

```bash
./cstool -d tc162 0 0xffffffff  # With details enabled
./cstool -d tc162 0 0xffffffff  # Without details
```

This way our chances for nasty surprises is lower.

**Update the new updater**

I replaced the `Update-Arch.sh` script with a proper Python scripts.
You need to add Alpha there and can drop the `Update-Arch.sh` change.

In `suite/auto-sync/Updater/` you need to:


- Add `Alpha` to `parse_args()` in `ASUpdater.py`, `CppTranslator.py` and `Differ.py`
- Add a configuration for Alpha in `suite/auto-sync/CppTranslator/arch_config.json`. The config file is documented in `CppTranslator/README.md`. 
- Run the [complete update procedure](https://github.com/Rot127/capstone/tree/as-docs/suite/auto-sync#update). It will fail at some point and you need to add some things to the Python scripts (e.g. to `CppTranslator/Patches/Includes.py`). Don't worry though, there shouldn't be many.

Also, I think the docs are still a confusingly written. And I update them in https://github.com/capstone-engine/capstone/pull/2217.

Please, if you have any suggestions or run against walls, let me know. So this stuff can be made easier.

@XVilka @kabeor This last point (updating the updater) could be also moved into a separated PR. Just in case it should be merged earlier.

**Comment by XVilka** [comment]
@kabeor could you please allow workflows to run again? Also, mark it as "ready for review"?

**Review by XVilka — APPROVED** [review]
LGTM as is. @R33v0LT Have you run the fuzzing?

**Comment by R33v0LT** [comment]
> LGTM as is. @R33v0LT Have you run the fuzzing?

In progress

**Comment by R33v0LT** [comment]
@XVilka, @Rot127, Fuzzing completed without fails

**Review by Rot127 — APPROVED** [review]
Please apply this patch:

```diff
diff --git a/suite/auto-sync/Updater/CppTranslator/Differ.py b/suite/auto-sync/Updater/CppTranslator/Differ.py
index 038e9829a..b3dd44f77 100755
--- a/suite/auto-sync/Updater/CppTranslator/Differ.py
+++ b/suite/auto-sync/Updater/CppTranslator/Differ.py
@@ -11,7 +11,7 @@ import difflib as dl
 import logging as log
 import sys
 
-from Configurator import Configurator
+from CppTranslator.Configurator import Configurator
 from Helper import (
     convert_loglevel,
     find_id_by_type,
```

Otherwise the updater throws an exception when run with the differ step.

Looks good to me otherwise.

**Comment by XVilka** [comment]
Ok, @kabeor @aquynh if you have no objections, this one is ready to merge too.

**Review by XVilka — APPROVED** [review]

**Review by Rot127 — APPROVED** [review]
@kabeor Please take a look when you find time.

**Comment by XVilka** [comment]
@kabeor @aquynh have you had a chance to look at it?

**Comment by kabeor** [comment]
It is really awesome, merged, thank you!

**Pull request merged**
